x-vector/src/X/Data/Vector/Grow.hs

```diff
@@ -0,0 +1,90 @@
+{-# LANGUAGE NoImplicitPrelude #-}
+module X.Data.Vector.Grow (
+    Grow
+  , new
+  , length
+  , capacity
+  , add
+  , clear
+  , freeze
+  , unsafeFreeze
+  , unsafeElems
+  ) where
+
+import           X.Data.Vector.Ref (Ref, newRef, readRef, modifyRef, modifyRefM)
+
+import           Control.Monad.Primitive (PrimMonad, PrimState)
+
+import           Data.Vector.Generic (Vector, Mutable)
+import qualified Data.Vector.Generic as Generic
+import           Data.Vector.Generic.Mutable (MVector)
+import qualified Data.Vector.Generic.Mutable as MGeneric
+import qualified Data.Vector.Mutable as MBoxed
+
+import           P hiding (length)
+
+
+data GrowState v s a =
+  GrowState {
+      growLength :: !Int
+    , growVector :: !(v s a)
+    }
+
+newtype Grow v s a =
+  Grow {
+      unGrow :: Ref MBoxed.MVector s (GrowState v s a)
+    }
+
+new :: (PrimMonad m, MVector v a) => Int -> m (Grow v (PrimState m) a)
+new n = do
+  gs <- GrowState 0 <$> MGeneric.new n
+  Grow <$> newRef gs
+{-# INLINE new #-}
+
+length :: PrimMonad m => Grow v (PrimState m) a -> m Int
+length =
+  fmap growLength . readRef . unGrow
+{-# INLINE length #-}
+
+capacity :: (PrimMonad m, MVector v a) => Grow v (PrimState m) a -> m Int
+capacity =
+  fmap (MGeneric.length . growVector) . readRef . unGrow
+{-# INLINE capacity #-}
+
+expand :: (PrimMonad m, MVector v a) => GrowState v (PrimState m) a -> m (GrowState v (PrimState m) a)
+expand gs@(GrowState n xs0) =
+  if n == MGeneric.length xs0 then
+    GrowState n <$> MGeneric.unsafeGrow xs0 (max n 1)
```

Comment:
@olorin the `unsafeGrow` function is just a bit surprising, it takes the number of elements to grow by as an argument (not the new size)

So here we grow by the current size, effectively doubling the size.